// @twurple/eventsub-base/es/events/chatNotifications/EventSubChannelChatCommunitySubGiftNotificationEvent.mjs
import { __decorate } from "tslib";
import { rawDataSymbol, rtfm } from '@twurple/common';
import { EventSubChannelChatBaseNotificationEvent } from "./EventSubChannelChatBaseNotificationEvent.mjs";
/**
 * An EventSub event representing a community sub gift notification in a channel's chat.
 */
let EventSubChannelChatCommunitySubGiftNotificationEvent = class EventSubChannelChatCommunitySubGiftNotificationEvent extends EventSubChannelChatBaseNotificationEvent {
    constructor() {
        super(...arguments);
        this.type = 'community_sub_gift';
    }
    /**
     * The ID of the community sub gift.
     */
    get id() {
        return this[rawDataSymbol].community_sub_gift.id;
    }
    /**
     * The tier of the subscriptions.
     */
    get tier() {
        return this[rawDataSymbol].community_sub_gift.sub_tier;
    }
    /**
     * The amount of gifts that are part of this community sub gift.
     */
    get amount() {
        return this[rawDataSymbol].community_sub_gift.total;
    }
    /**
     * The amount of gifts that the gifter has sent in total, or `null` the gift is anonymous.
     */
    get cumulativeAmount() {
        return this[rawDataSymbol].community_sub_gift.cumulative_total;
    }
};
EventSubChannelChatCommunitySubGiftNotificationEvent = __decorate([
    rtfm('eventsub-base', 'EventSubChannelChatCommunitySubGiftNotificationEvent', 'broadcasterId')
], EventSubChannelChatCommunitySubGiftNotificationEvent);
export { EventSubChannelChatCommunitySubGiftNotificationEvent };
