// @twurple/easy-bot/lib/events/RaidCancelEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RaidCancelEvent = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const chat_1 = require("@twurple/chat");
const common_1 = require("@twurple/common");
/**
 * An event representing an outgoing raid being canceled.
 *
 * @meta category events
 */
let RaidCancelEvent = class RaidCancelEvent {
    /** @internal */
    constructor(channel, msg, bot) {
        this._broadcasterName = (0, chat_1.toUserName)(channel);
        this._msg = msg;
        this._bot = bot;
    }
    /**
     * The ID of the broadcaster.
     */
    get broadcasterId() {
        return this._msg.channelId;
    }
    /**
     * The name of the broadcaster.
     */
    get broadcasterName() {
        return this._broadcasterName;
    }
    /**
     * Gets more information about the broadcaster.
     */
    async getBroadcaster() {
        return (0, common_1.checkRelationAssertion)(await this._bot.api.users.getUserById(this.broadcasterId));
    }
    /**
     * The ID of the user who canceled the raid.
     */
    get userId() {
        return this._msg.userInfo.userId;
    }
    /**
     * The name of the user who canceled the raid.
     */
    get userName() {
        return this._msg.userInfo.userName;
    }
    /**
     * The display name of the user who canceled the raid.
     */
    get userDisplayName() {
        return this._msg.userInfo.displayName;
    }
    /**
     * Gets more information about the user who canceled the raid.
     */
    async getUser() {
        return (0, common_1.checkRelationAssertion)(await this._bot.api.users.getUserById(this.userId));
    }
    /**
     * The full object that contains all the message information.
     */
    get messageObject() {
        return this._msg;
    }
};
exports.RaidCancelEvent = RaidCancelEvent;
tslib_1.__decorate([
    (0, shared_utils_1.Enumerable)(false)
], RaidCancelEvent.prototype, "_broadcasterName", void 0);
tslib_1.__decorate([
    (0, shared_utils_1.Enumerable)(false)
], RaidCancelEvent.prototype, "_msg", void 0);
tslib_1.__decorate([
    (0, shared_utils_1.Enumerable)(false)
], RaidCancelEvent.prototype, "_bot", void 0);
exports.RaidCancelEvent = RaidCancelEvent = tslib_1.__decorate([
    (0, common_1.rtfm)('easy-bot', 'RaidCancelEvent', 'broadcasterName')
], RaidCancelEvent);
