// @twurple/easy-bot/lib/events/SubsOnlyToggleEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubsOnlyToggleEvent = void 0;
const tslib_1 = require("tslib");
const shared_utils_1 = require("@d-fischer/shared-utils");
const chat_1 = require("@twurple/chat");
const common_1 = require("@twurple/common");
/**
 * An event representing subscriber-only mode being toggled in a channel.
 *
 * @meta category events
 */
let SubsOnlyToggleEvent = class SubsOnlyToggleEvent {
    /** @internal */
    constructor(channel, enabled, bot) {
        this._broadcasterName = (0, chat_1.toUserName)(channel);
        this._enabled = enabled;
        this._bot = bot;
    }
    /**
     * The name of the broadcaster.
     */
    get broadcasterName() {
        return this._broadcasterName;
    }
    /**
     * Gets more information about the broadcaster.
     */
    async getBroadcaster() {
        return (0, common_1.checkRelationAssertion)(await this._bot.api.users.getUserByName(this.broadcasterName));
    }
    /**
     * Whether subscriber-only mode was enabled.
     *
     * `true` means it was enabled, `false` means it was disabled.
     */
    get enabled() {
        return this._enabled;
    }
    /**
     * Enables subscriber-only mode in the channel.
     */
    async enable() {
        await this._bot.enableSubsOnly(this.broadcasterName);
    }
    /**
     * Disables subscriber-only mode in the channel.
     */
    async disable() {
        await this._bot.disableSubsOnly(this.broadcasterName);
    }
};
exports.SubsOnlyToggleEvent = SubsOnlyToggleEvent;
tslib_1.__decorate([
    (0, shared_utils_1.Enumerable)(false)
], SubsOnlyToggleEvent.prototype, "_broadcasterName", void 0);
tslib_1.__decorate([
    (0, shared_utils_1.Enumerable)(false)
], SubsOnlyToggleEvent.prototype, "_enabled", void 0);
tslib_1.__decorate([
    (0, shared_utils_1.Enumerable)(false)
], SubsOnlyToggleEvent.prototype, "_bot", void 0);
exports.SubsOnlyToggleEvent = SubsOnlyToggleEvent = tslib_1.__decorate([
    (0, common_1.rtfm)('easy-bot', 'SubsOnlyToggleEvent', 'broadcasterName')
], SubsOnlyToggleEvent);
