// @twurple/eventsub-base/es/events/EventSubChannelChatSettingsUpdateEvent.mjs
import { __decorate } from "tslib";
import { Enumerable } from '@d-fischer/shared-utils';
import { checkRelationAssertion, DataObject, rawDataSymbol, rtfm } from '@twurple/common';
/**
 * An EventSub event representing chat settings being updated in a channel.
 */
let EventSubChannelChatSettingsUpdateEvent = class EventSubChannelChatSettingsUpdateEvent extends DataObject {
    /** @internal */
    constructor(data, client) {
        super(data);
        this._client = client;
    }
    /**
     * The ID of the broadcaster.
     */
    get broadcasterId() {
        return this[rawDataSymbol].broadcaster_user_id;
    }
    /**
     * The name of the broadcaster.
     */
    get broadcasterName() {
        return this[rawDataSymbol].broadcaster_user_login;
    }
    /**
     * The display name of the broadcaster.
     */
    get broadcasterDisplayName() {
        return this[rawDataSymbol].broadcaster_user_name;
    }
    /**
     * Gets more information about the broadcaster.
     */
    async getBroadcaster() {
        return checkRelationAssertion(await this._client.users.getUserById(this[rawDataSymbol].broadcaster_user_id));
    }
    /**
     * Whether emote only mode is enabled.
     */
    get emoteOnlyModeEnabled() {
        return this[rawDataSymbol].emote_mode;
    }
    /**
     * Whether follower only mode is enabled.
     */
    get followerOnlyModeEnabled() {
        return this[rawDataSymbol].follower_mode;
    }
    /**
     * The time after which users are able to send messages after following, in minutes.
     *
     * Is `null` if follower only mode is not enabled,
     * but may also be `0` if you can send messages immediately after following.
     */
    get followerOnlyModeDelay() {
        return this[rawDataSymbol].follower_mode_duration_minutes;
    }
    /**
     * Whether slow mode is enabled.
     */
    get slowModeEnabled() {
        return this[rawDataSymbol].slow_mode;
    }
    /**
     * The time to wait between messages in slow mode, in seconds.
     *
     * Is `null` if slow mode is not enabled.
     */
    get slowModeDelay() {
        return this[rawDataSymbol].slow_mode_wait_time_seconds;
    }
    /**
     * Whether subscriber only mode is enabled.
     */
    get subscriberOnlyModeEnabled() {
        return this[rawDataSymbol].subscriber_mode;
    }
    /**
     * Whether unique chat mode is enabled.
     */
    get uniqueChatModeEnabled() {
        return this[rawDataSymbol].unique_chat_mode;
    }
};
__decorate([
    Enumerable(false)
], EventSubChannelChatSettingsUpdateEvent.prototype, "_client", void 0);
EventSubChannelChatSettingsUpdateEvent = __decorate([
    rtfm('eventsub-base', 'EventSubChannelChatSettingsUpdateEvent', 'broadcasterId')
], EventSubChannelChatSettingsUpdateEvent);
export { EventSubChannelChatSettingsUpdateEvent };
