// tubechat/lib/lib/ChatMessages.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchChat = void 0;
function fetchChat(continuation, apiKey, clientName, clientVersion) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const bodyPost = {
                context: {
                    client: {
                        clientName,
                        clientVersion
                    }
                },
                continuation
            };
            const requestOptions = {
                method: 'POST',
                headers: {
                    'content-type': 'application/json'
                },
                body: JSON.stringify(bodyPost),
            };
            const fetchMessages = yield fetch(`https://www.youtube.com/youtubei/v1/live_chat/get_live_chat?key=${apiKey}&prettyPrint=false`, requestOptions);
            const messageData = yield fetchMessages.json();
            return messageData;
        }
        catch (err) {
            return err;
        }
    });
}
exports.fetchChat = fetchChat;
