// @twurple/pubsub/es/messages/PubSubChatModActionMessage.mjs
import { __decorate } from "tslib";
import { DataObject, rawDataSymbol, rtfm } from '@twurple/common';
/**
 * A message that informs about a moderation action being performed in a channel.
 */
let PubSubChatModActionMessage = class PubSubChatModActionMessage extends DataObject {
    /** @internal */
    constructor(data, _channelId) {
        super(data);
        this._channelId = _channelId;
    }
    /**
     * The ID of the channel where the action was performed.
     */
    get channelId() {
        return this._channelId;
    }
    /**
     * The type of the message.
     */
    get type() {
        return this[rawDataSymbol].data.type;
    }
    /**
     * The action that was performed.
     */
    get action() {
        return this[rawDataSymbol].data.moderation_action;
    }
    /**
     * The arguments given to the action.
     */
    get args() {
        return this[rawDataSymbol].data.args;
    }
    /**
     * The user ID of the moderator that performed the action.
     */
    get userId() {
        return this[rawDataSymbol].data.created_by_user_id;
    }
    /**
     * The name of the moderator that performed the action.
     */
    get userName() {
        return this[rawDataSymbol].data.created_by;
    }
};
PubSubChatModActionMessage = __decorate([
    rtfm('pubsub', 'PubSubChatModActionMessage', 'userId')
], PubSubChatModActionMessage);
export { PubSubChatModActionMessage };
