// @twurple/pubsub/es/messages/PubSubAutoModQueueMessage.mjs
import { __decorate } from "tslib";
import { DataObject, rawDataSymbol, rtfm } from '@twurple/common';
/**
 * A message that informs about a message being processed in the AutoMod queue.
 */
let PubSubAutoModQueueMessage = class PubSubAutoModQueueMessage extends DataObject {
    /** @internal */
    constructor(data, _channelId) {
        super(data);
        this._channelId = _channelId;
    }
    /**
     * The ID of the channel where the message was posted.
     */
    get channelId() {
        return this._channelId;
    }
    /**
     * The ID of the message.
     */
    get messageId() {
        return this[rawDataSymbol].data.message.id;
    }
    /**
     * The content of the message.
     */
    get messageContent() {
        return this[rawDataSymbol].data.message.content.text;
    }
    /**
     * The fragments of the message that were found to be against the moderation level of the channel.
     */
    get foundMessageFragments() {
        return this[rawDataSymbol].data.message.content.fragments;
    }
    /**
     * The ID of the user that sent the message.
     */
    get senderId() {
        return this[rawDataSymbol].data.message.sender.user_id;
    }
    /**
     * The name of the user that sent the message.
     */
    get senderName() {
        return this[rawDataSymbol].data.message.sender.login;
    }
    /**
     * The display name of the user that sent the message.
     */
    get senderDisplayName() {
        return this[rawDataSymbol].data.message.sender.display_name;
    }
    /**
     * The chat color of the user that sent the message.
     */
    get senderColor() {
        return this[rawDataSymbol].data.message.sender.chat_color;
    }
    /**
     * The date when the message was sent.
     */
    get sendDate() {
        return new Date(this[rawDataSymbol].data.message.sent_at);
    }
    /**
     * The classification of the message content.
     */
    get contentClassification() {
        return this[rawDataSymbol].data.content_classification;
    }
    /**
     * The status of the queue entry.
     */
    get status() {
        return this[rawDataSymbol].data.status;
    }
    /**
     * The ID of the user that resolved the queue entry, or null if it was not resolved or timed out.
     */
    get resolverId() {
        return this[rawDataSymbol].data.resolver_id || null;
    }
    /**
     * The name of the user that resolved the queue entry, or null if it was not resolved or timed out.
     */
    get resolverName() {
        return this[rawDataSymbol].data.resolver_login || null;
    }
};
PubSubAutoModQueueMessage = __decorate([
    rtfm('pubsub', 'PubSubAutoModQueueMessage', 'messageId')
], PubSubAutoModQueueMessage);
export { PubSubAutoModQueueMessage };
